### PR TITLE
Feature/spectrum reference

### PR DESCRIPTION
**Pull request by mwalzer**
added spectrum reference annotation for idXML in IDFileConverter, if input is idXML (most likely, these got lost during reading of the original fileformat of the searchengine, if it was even annotated)

also added a function to add do optional ion matching in SpectraMetaDataLookup when spectra are matched anyways  for  spectrum reference lookup

### CODE REVIEW

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h [hunk 1]
```diff
@@ -267,9 +267,24 @@ namespace OpenMS
 
        The raw data is only loaded from @p filename if necessary, i.e. if there are any peptide IDs with missing RTs.
     */
-    static bool addMissingRTsToPeptideIDs(
-      std::vector<PeptideIdentification>& peptides, const String& filename,
-      bool stop_on_error = false);
+    static bool addMissingRTsToPeptideIDs(std::vector<PeptideIdentification>& peptides, const String& filename,
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: please add parameter documentation

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h [hunk 2]
```diff
@@ -267,9 +267,24 @@ namespace OpenMS
...
+       @param filename Name of a raw data file (e.g. mzML) for looking up spectrum_references (nativeID)
+       @param stop_on_error Stop when an ID could not be matched to a spectrum (or keep going)?
+
+       @return True if all peptide IDs could be annotated successfully (including if all already had spectrum_reference values), false otherwise.
+
+       Look-up works by matching the RT of a peptide ID to the RT of a spectrum. Only peptide IDs without spectrum_reference are looked up; the new meta value is set to the native ID of the corresponding spectrum.
+
+       The raw data is only loaded from @p filename if necessary, i.e. if there are any peptide IDs with spectrum_references.
+    */
+    static bool addMissingSpectrumReferencestoPeptideIDs(std::vector<PeptideIdentification>& peptides, const String& filename,
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: param doc

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h [hunk 3]
```diff
@@ -283,6 +298,8 @@ namespace OpenMS
     /// Assignment operator (not implemented)
     SpectrumMetaDataLookup& operator=(const SpectrumMetaDataLookup&);
 
+    /// adds Ion matches to identifcations
+    static void addIonMatches_(PeptideIdentification& pi, const MSSpectrum<Peak1D> &spec);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: this is not lookup but reannotation with our algorithms. @hendrikweisser would you place this in this class?

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp [hunk 4]
```diff
@@ -177,17 +183,21 @@ namespace OpenMS
...
-
   bool SpectrumMetaDataLookup::addMissingRTsToPeptideIDs(
     vector<PeptideIdentification>& peptides, const String& filename,
-    bool stop_on_error)
+    bool stop_on_error, bool reset_basename)
   {
     MSExperiment<> exp;
     SpectrumLookup lookup;
     bool success = true;
+    String bn = basename(filename.c_str());
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: see CDASH

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp [hunk 5]
```diff
@@ -212,4 +222,254 @@ namespace OpenMS
...
+            success = false;
+            if (stop_on_error) break;
+          }
+        }
+      }
+    }
+    return success;
+  }
+
+  void SpectrumMetaDataLookup::addIonMatches_(PeptideIdentification& pi, const MSSpectrum<Peak1D>& spec)
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: While it is nice to have the possibility of reannotation I would not include it in its current form.
Maybe you can give some details on what you are intending with this function?

File: src/topp/IDFileConverter.cpp [hunk 6]
```diff
@@ -327,12 +329,29 @@ class TOPPIDFileConverter :
           MascotXMLFile::initializeLookup(lookup, exp, scan_regex);
           PepXMLFile().load(in, protein_identifications,
                             peptide_identifications, exp_name, lookup);
+          bool reset_basename = getFlag_("reset_basename");
+          bool add_ions = getFlag_("add_ionmatch_annotation");
+          if (add_ions)
+          {
+            SpectrumMetaDataLookup::addMissingSpectrumReferencestoPeptideIDs(
+                      peptide_identifications, exp_name, false, add_ions, reset_basename);
+          }
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: here is probably the right position to do MASCOT specific reannotation (e.g. you know the parameters and ion types here)

File: src/topp/IDFileConverter.cpp [hunk 7]
```diff
@@ -327,12 +329,29 @@ class TOPPIDFileConverter :
...
         IdXMLFile().load(in, protein_identifications, peptide_identifications);
+        // get native id from mz_file:
+        String exp_name = getStringOption_("mz_file");
+        if (!exp_name.empty())
+        {
+          bool add_ions = getFlag_("add_ionmatch_annotation");
+          bool reset_basename = getFlag_("reset_basename");
+          SpectrumMetaDataLookup::addMissingSpectrumReferencestoPeptideIDs(
+                      peptide_identifications, exp_name, false, add_ions, reset_basename);
+        }
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: or here

File: src/topp/IDFileConverter.cpp [hunk 8]
```diff
@@ -343,10 +362,17 @@ class TOPPIDFileConverter :
...
         if (!exp_name.empty())
         {
           SpectrumMetaDataLookup::addMissingRTsToPeptideIDs(
-            peptide_identifications, exp_name, false);
+            peptide_identifications, exp_name, false, reset_basename);
+          bool add_ions = getFlag_("add_ionmatch_annotation");
+          if (add_ions)
+          {
+            SpectrumMetaDataLookup::addMissingSpectrumReferencestoPeptideIDs(
+                      peptide_identifications, exp_name, false, add_ions, reset_basename);
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: or here

File: src/topp/IDMerger.cpp [hunk 9]
```diff
@@ -267,10 +267,11 @@ class TOPPIDMerger :
     else
     {
       bool annotate_file_origin = getFlag_("annotate_file_origin");
-      map<String, ProteinIdentification> proteins_by_id;
+      map<String, ProteinIdentification> proteins_by_id; //id<String> the reference which links PeptideIdentification and ProteinIdentification (must be equal)
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: space after //

Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: what does id<String> mean? is this the format?

**Comment by jpfeuffer** [comment]
This PR sounds very interesting. What do you need it for? How would an additionally annotated output file look like?
Does this mean I could save the peak matches in an ID file? Would be helpful for visualization.

File: src/topp/IDMerger.cpp [hunk 10]
```diff
@@ -267,10 +267,11 @@ class TOPPIDMerger :
...
       vector<vector<PeptideIdentification> > peptides_by_file;
       StringList add_to_ids; // IDs from the "add_to" file (if any)
 
+      ///make sure add_to at front of the file_names and is not duplicated
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: why is adding to front important?

File: src/topp/IDMerger.cpp [hunk 11]
```diff
@@ -278,6 +279,8 @@ class TOPPIDMerger :
       }
 
       peptides_by_file.resize(file_names.size());
+      ///reassign ProteinIdentificantion.id if not unique, update all
+      ///corresponding PeptideIdentification.id (links)
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: spaces

File: src/topp/IDMerger.cpp [hunk 12]
```diff
@@ -321,7 +324,9 @@ class TOPPIDMerger :
         }
       }
 
-      if (add_to.empty()) // copy proteins from map into vector for writing
+      /// copy all ProteinIdentification and PeptideIdentification from each file
+      /// into newfrom map into vector for writing
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: newfrom

File: src/topp/IDMerger.cpp [hunk 13]
```diff
@@ -335,16 +340,18 @@ class TOPPIDMerger :
...
-        // copy over data from reference file ("add_to"):
+        ///select ProteinIdentifications from reference files ("add_to")
         map<String, ProteinIdentification> selected_proteins;
         for (StringList::iterator ids_it = add_to_ids.begin();
              ids_it != add_to_ids.end(); ++ids_it)
         {
           selected_proteins[*ids_it] = proteins_by_id[*ids_it];
         }
-        // keep track of peptides that shouldn't be duplicated:
+        ///set up set of peptide sequences from only the first reference file ("add_to")
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: space

File: src/topp/IDMerger.cpp [hunk 14]
```diff
@@ -335,16 +340,18 @@ class TOPPIDMerger :
...
+        ///which will be ignored from here on (coming from all other files)
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: space

File: src/topp/IDMerger.cpp [hunk 15]
```diff
@@ -356,7 +363,9 @@ class TOPPIDMerger :
         }
         peptides.insert(peptides.end(), base_peptides.begin(),
                         base_peptides.end());
-        // merge in data from other files:
+        ///cut all other PeptideIdentifications to the first hit after sorting
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: spaces

File: src/topp/IDMerger.cpp (hunk 14 above)
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: I don't get the comment

File: src/topp/IDMerger.cpp [hunk 16]
```diff
@@ -356,7 +363,9 @@ class TOPPIDMerger :
...
+        /// and add new sequences, add the first ProteinHits corresponding to
+        /// the new sequence
```
Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: comment is hard to follow

Comment (mwalzer) [inline src/topp/IDMerger.cpp]: so "// merge in data from other files:" was better??

Comment (timosachsenberg) [inline src/topp/IDMerger.cpp]: no but not much worse (at least regarding my understanding on what is happening in the code)

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h (hunk 3 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: _re_ annotation?? I think this is not prevalent information in id. Also, it is lookup, as the data has to be gathered from the spectra data.

File: src/topp/IDFileConverter.cpp (hunk 6 above)
Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: what do you mean?

File: src/topp/IDFileConverter.cpp (hunk 7 above)
Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: what do you mean?

File: src/topp/IDFileConverter.cpp (hunk 8 above)
Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: what do you mean?

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h (hunk 3 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: This information might well be present from the search engine results as spectra have been potentially already been annotated using custom algorithms. But we might not carry over the information.
It is cleary no lookup as no existing information is looked up but recalculated.

File: src/topp/IDFileConverter.cpp (hunk 6 above)
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: reannotation without knowing details on fragmentation method / tolerances does make very litte sense. Here you would have this information available and could call e.g. a method from a not-ecisting SpectrumReannotation .class.

File: src/topp/IDMerger.cpp (hunk 16 above)
Comment (mwalzer) [inline src/topp/IDMerger.cpp]: Basically, adding IDMerger to the PR was rather an accident. I was looking at the tool and to better understand it's workings i thought I could add comments.

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp (hunk 5 above)
Comment (mwalzer) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: Why would you make acceptance dependent on the use?
As you already said, it would be nice to have it. To be able to look up what parts of the spectrum can be explained by the identification. This will be crucial when you do 'closequarters' ID-QC.

Comment (timosachsenberg) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: I will accept it happily if it is moved to a separate class for spectrum annotation. Here it just doesn't make sense as the important information like tolerances, fragmentation type etc. is not provided. The information will just do more harm then good if e.g. you annotate a CID spectrum with HCD settings or vice versa.

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp (hunk 4 above)
Comment (mwalzer) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: Where?

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp (hunk 5 above)
Comment (mwalzer) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: Yeah, I see that. I put it here because lookup is pretty time expensive itself by reading the spectra data and sifting through them. Being able to do it on the go as you anyways got through while spectrum_reference annotation.

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h (hunk 3 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: :wink: yeah in a perfect world, maybe. No, MSGF+ is the **_only_** one to report such thing.
It clearly is lookup as the spectra have to be looked up. What the class is doing.

File: src/openms/source/METADATA/SpectrumMetaDataLookup.cpp (hunk 5 above)
Comment (mwalzer) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: So, where should I put it without the overhead of going through the spectra data again?

Comment (timosachsenberg) [inline src/openms/source/METADATA/SpectrumMetaDataLookup.cpp]: Not if you do it after reannotation.
Anyways, let's discuss this offline

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h (hunk 3 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: I don't say that you can't use this class for lookup. I just say this class is a bad place to put the ion annotation code. But as I said - lets discuss this offline.

**Comment by mwalzer** [comment]
@jpfeuffer right now it would be a userParam
```<UserParam type="string" name="matched_ions" value="y1+,a4++-H2O1,a4++,b2+-H3N1,y2+,b5++-H2O1,x5++,c6++-H3N1,c3+-H2O1,y6++-H3N1,y6++,x6++,z7++-H2O1,a7++-H3N1,y3+,a7++,b7++-H2O1,y7++,x7++-H2O1,c7++-H3N1,x3+,x7++,a8++-H3N1,a8++,a4+,z8++-H2O1,c8++-H3N1,b4+-H3N1,c8++,y8++,x8++-H3N1,c4+,x4+,a5+-H2O1,a5+-H3N1,a5+,b5+-H2O1,c5+-H2O1,c5+,a6+-H2O1,a6+,b6+-H2O1,b6+-H3N1,c6+-H3N1,z6+,y6+,x6+,a7+-H3N1,b7+-H2O1,b7+,a8+-H2O1,c8+-H2O1,y8+-H3N1,y8+"/>````
But it really bloats the file, but I'd need it for a closer look at quality and to get an idea of how sound the footing of a identification is.

**Comment by mwalzer** [comment]
Okay, so I will pull it out there into a SpectrumAnnotator class

**Comment by timosachsenberg** [comment]
perfect - I am sure this will be used in many other places including visualization and annotation in e.g. mzTab files

**Comment by timosachsenberg** [comment]
maybe you forgot some changes: e.g. http://cdash.openms.de/testDetails.php?test=30171807&build=163065

**Comment by mwalzer** [comment]
Just added more items to the stats, refactoring is still pending - no worries.

File: src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h (hunk 1 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: I think there is nothing missing anymore

Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/SpectrumMetaDataLookup.h]: +1

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+namespace OpenMS
+{
+  const boost::regex SpectrumAnnotator::nt_regex_("[a,b,c][[:digit:]]+[+]+");
+  const boost::regex SpectrumAnnotator::ct_regex_("[x,y,z][[:digit:]]+[+]+");
+  const boost::regex SpectrumAnnotator::noloss_regex_("[a,b,c,x,y,z][[:digit:]]+[+]+");
+
+  SpectrumAnnotator::SpectrumAnnotator() :
+    DefaultParamHandler("SpectrumAnnotator")
+  {
+    defaults_.setValue("basic_statistics", "true", "If set, meta values for peak_number, sum_intensity, matched_ion_number, matched_intensity are added");
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: another possibility that greatly reduces the number of parameters would be to only provide one StringList that contains all flags.
But fine with me if you prefer it that way.

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+  }
+
+  SpectrumAnnotator::~SpectrumAnnotator()
+  {
+  }
+
+  MSSpectrum<RichPeak1D> SpectrumAnnotator::annotateMatches(const PeptideHit& ph, const MSSpectrum<Peak1D>& spec, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    MSSpectrum<RichPeak1D> rich_theoretical_spec;
+    MSSpectrum<RichPeak1D> rich_input_spectrum = MSSpectrumHelper::clone(spec);
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: instead of the clone function I would prefer a spcialized constructor in MSSpectrum<RichPeak1D>:
template<> MSSpectrum<RichPeak1D>::MSSpectrum(MSSpectrum<Peak1D>){...}

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+    }
+    sa.getSpectrumAlignment(al, rich_theoretical_spec, rich_input_spectrum); //peaks from theor. may be matched to none or one in spec!
+    for (vector<pair<Size, Size > >::const_iterator it = al.begin(); it != al.end(); ++it)
+    {
+        rich_input_spectrum[it->second].setMetaValue("IonMatchError", std::fabs(rich_input_spectrum[it->second].getMZ() - rich_theoretical_spec[it->first].getMZ()));
+        rich_input_spectrum[it->second].setMetaValue("IonName", rich_theoretical_spec[it->first].getMetaValue("IonName"));
+    }
+    Param sap = sa.getParameters();
+    rich_input_spectrum.setMetaValue("fragment_match_tolerance", sap.getValue("tolerance"));
+    return rich_input_spectrum;
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: you could also set the unit.
By the way, we previously used: fragment_mass_tolerance and fragment_mass_tolerance_ppm so maybe we can reuse these meta value names

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+        mzs.reserve(rich_spec.size());
+
+        double nint = 0;
+        double cint = 0;
+
+        StringList allowed_types = ListUtils::create<String>("y,b,a,c,x,z");
+        map<String, vector<bool> > ion_series;
+        for (StringList::iterator st = allowed_types.begin(); st != allowed_types.end(); ++st)
+        {
+          ion_series.insert(make_pair(*st, vector<bool>(ph->getSequence().size()-1, false)));
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: why size() -1 ?

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+        for (RichPeakSpectrum::Iterator it = rich_spec.begin(); it != rich_spec.end(); ++it)
+        {
+          sum_intensity += it->getIntensity();
+          if (it->metaValueExists("IonMatchError"))
+          {
+            fragmenterrors.push_back(it->getMetaValue("IonMatchError"));
+            intensities.push_back(it->getIntensity());
+            match_intensity += it->getIntensity();
+            mzs.push_back(it->getMZ());
+            String ion_name = it->getMetaValue("IonName");
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: consider to use more: const String& where possible for improved performance

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+              }
+              if (max_series_ && boost::regex_match(ion_name, noloss_regex_))
+              {
+                String ion_type = ion_name.prefix(1);
+                if (ListUtils::contains(allowed_types, ion_type)) //case sensitive?
+                {
+                  try
+                  {
+                    int i = ion_name.substr(1).remove('+').toInt() - 1; //do not care about charge
+                    ion_series[ion_type].at(i) = true;
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: charge related features might be worth looking at in the future

File: src/topp/IDFileConverter.cpp [hunk 23]
```diff
@@ -170,6 +176,9 @@ class TOPPIDFileConverter :
     registerFlag_("ignore_proteins_per_peptide", "[Sequest only] Workaround to deal with .out files that contain e.g. \"+1\" in references column,\n"
                                                  "but do not list extra references in subsequent lines (try -debug 3 or 4)", true);
     registerStringOption_("scan_regex", "<expression>", "", "[Mascot, pepXML, Percolator only] Regular expression used to extract the scan number or retention time. See documentation for details.", false, true);
+    registerFlag_("add_spectrum_references", "Adds the spectrum files spectrum ID reference to each PeptideIdentification. Warning: works only with mz_file set", true);
+    registerDoubleOption_("add_ionmatch_annotation", "<tolerance value>", 0.5, "Adds IonMatch statistics to each PeptideHit if tolerance value > 0. Warning: works only with mz_file set, might be time and memory laborious!", false, true);
+    registerFlag_("reset_basename", "Resets the BaseName of every PeptideIdentification with the basename of the given mzML file.", true);
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: hmm I am not sure if setting this  in the PI this is sufficient.
I would also call this spectra_data as in ProteinIdentification/Feature and ConsensusFeature.
In the file name list in ProteinIdentification it might need to be updated to.

**Comment by timosachsenberg** [comment]
In general it looks very good and is a nice addition. See some minor comments and (maybe the biggest item: the last comment about the spectra_data meta value (which is already in use in OpenMS))

File: src/topp/IDFileConverter.cpp [hunk 24]
```diff
@@ -170,6 +176,9 @@ class TOPPIDFileConverter :
     registerFlag_("ignore_proteins_per_peptide", "[Sequest only] Workaround to deal with .out files that contain e.g. \"+1\" in references column,\n"
                                                  "but do not list extra references in subsequent lines (try -debug 3 or 4)", true);
     registerStringOption_("scan_regex", "<expression>", "", "[Mascot, pepXML, Percolator only] Regular expression used to extract the scan number or retention time. See documentation for details.", false, true);
+    registerFlag_("add_spectrum_references", "Adds the spectrum files spectrum ID reference to each PeptideIdentification. Warning: works only with mz_file set", true);
```
Comment (hendrikweisser) [inline src/topp/IDFileConverter.cpp]: Better: "Adds a reference to the spectrum file's spectrum ID to each peptide identification. This requires 'mz_file' to be set."
"PeptideIdentification" is the name of the OpenMS class, which is irrelevant for users. The actual entity is a "peptide identification"!

File: src/topp/IDFileConverter.cpp [hunk 25]
```diff
@@ -170,6 +176,9 @@ class TOPPIDFileConverter :
...
+    registerDoubleOption_("add_ionmatch_annotation", "<tolerance value in Da>", 0, "Adds IonMatch statistics to each PeptideHit if tolerance value > 0 e.g. 0.5. Warning: works only with mz_file set, might be time and memory laborious!", false, true);
```
Comment (hendrikweisser) [inline src/topp/IDFileConverter.cpp]: Better: "Tolerance in Da for ion match annotation. If set to a positive value, ion match statistics are added to each peptide hit. This requires 'mz_file' to be set, and might be time/memory intensive."
Again: "PeptideHit" -> OpenMS class; "peptide hit" -> actual entity

File: src/topp/IDFileConverter.cpp [hunk 26]
```diff
@@ -170,6 +176,9 @@ class TOPPIDFileConverter :
...
+    registerFlag_("reset_basename", "Resets the BaseName of every PeptideIdentification with the basename of the given mzML file.", true);
```
Comment (hendrikweisser) [inline src/topp/IDFileConverter.cpp]: Better: "Sets the 'basename' attribute of every peptide identification to the base name of the mzML file given via 'mz_file'."

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 17 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: Yeah, even if it is not _totally_ consistent within openms I found the list of flags is more used for the tools and the params more used for the algo classes. But thats just a wibblywobbly valuation of mine. Still, I would stick to the params.

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 18 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: tried. Didn't get it to work. After merge, feel free to try yourself, I'd like a specialised constructor solution better as well. One problem though is, that you have to specialise the whole class in that case...

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 19 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: true, willfix

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 22 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: Yeah, there is a lot of interesting stuff to look at. I'll add a TODO from your suggestion

File: src/topp/IDFileConverter.cpp (hunk 23 above)
Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: well, that feature was there before if I recall right, and was dragged over during all that interface change and refactoring. This is an IDFileConverter specific issue and I'd prefer it to be resolved in a separate issue/PR

**Comment by timosachsenberg** [comment]
@hendrikweisser what is the purpose of basename again? Maybe we can come up with something that is compatible with our spectra_data approach of encoding the names of the original mzML files (which is internally stored and accessible via primaryMSRun)

File: src/topp/IDFileConverter.cpp (hunk 24 above)
Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: Yes, thanks!!

**Comment by mwalzer** [comment]
Yeah, I guess it would be helpful if we had a mechanism to add 'spectra_data' where it is missing/got lost.

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 20 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: TheoreticalSpectrumGenerator does no such thing, or does it? https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L356
What would your suggestion be?

**Comment by hendrikweisser** [comment]
> what is the purpose of basename again?

Contrary to what @mwalzer said, that flag ("reset_basename") was *not* present in IDFileConverter before - see the current code: https://github.com/OpenMS/OpenMS/blob/develop/src/topp/IDFileConverter.cpp

The basename plays in role in reading/writing pepXML, so maybe that's why it was added?

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 20 above)
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: TSG it would not crash if e.g. a0 or a_size are generated but your code might - so maybe adding at leat a comment which ions are considered is a good idea for upcoming bug hunters ;)

**Comment by timosachsenberg** [comment]
what is the state of the PR?

**Comment by timosachsenberg** [comment]
tests are missing as far as I can see

**Comment by mwalzer** [comment]
And I think @hendrikweisser had some PRs in between that were touching the spectrum reference handling? I'd have to check for conflicts too.

**Comment by mwalzer** [comment]
branch suffered scope creep and incompatible/same effect parallel development. Will close.
